internal/driver/readhandler.go

```diff
@@ -24,23 +46,86 @@ func (d *Driver) HandleReadCommands(deviceName string, protocols map[string]mode
 
 	d.Logger.Debugf("Driver.HandleReadCommands: protocols: %v resource: %v attributes: %v", protocols, reqs[0].DeviceResourceName, reqs[0].Attributes)
 
-	// create device client and open connection
-	endpoint, err := config.FetchEndpoint(protocols)
+	availableServerEndpoints, err := opcua.GetEndpoints(d.serviceConfig.OPCUAServer.Endpoint)
 	if err != nil {
-		return nil, err
+		d.Logger.Error("OPC GetEndpoints: %w", err)
+	}
+	credentials, err := getCredentials(d.serviceConfig.OPCUAServer.CredentialsPath)
+	if err != nil {
+		d.Logger.Error("getCredentials: %w", err)
+	}
+
+	username := credentials.Username
+	password := credentials.Password
+	policy := ua.SecurityPolicyURIBasic256Sha256
+	mode := ua.MessageSecurityModeSignAndEncrypt
+
+	ep := opcua.SelectEndpoint(availableServerEndpoints, policy, mode)
+	c, err := generateCert() // This is where you generate the certificate
+	if err != nil {
+		d.Logger.Error("generateCert: %w", err)
 	}
-	ctx := context.Background()
 
-	client := opcua.NewClient(endpoint, opcua.SecurityMode(ua.MessageSecurityModeNone))
+	pk, ok := c.PrivateKey.(*rsa.PrivateKey) // This is where you set the private key
+	if !ok {
+		log.Print("invalid private key")
+	}
+
+	cert := c.Certificate[0]
+
+	opts := []opcua.Option{
+		opcua.SecurityPolicy(policy),
+		opcua.SecurityMode(mode),
+		opcua.PrivateKey(pk),
+		opcua.Certificate(cert),                // Set the certificate for the OPC UA Client
+		opcua.AuthUsername(username, password), // Use this if you are using username and password
+		opcua.SecurityFromEndpoint(ep, ua.UserTokenTypeUserName),
+		opcua.SessionTimeout(30 * time.Minute),
+	}
+
+	ctx := context.Background()
+	client := opcua.NewClient(d.serviceConfig.OPCUAServer.Endpoint, opts...)
 	if err := client.Connect(ctx); err != nil {
-		d.Logger.Warnf("Driver.HandleReadCommands: Failed to connect OPCUA client, %s", err)
-		return nil, err
+		d.Logger.Error("Driver.HandleReadCommands: Failed to connect OPCUA client, %s", err)
```

Comment:
Same as above. Here it could make sense also to leave it a warning because it could be no error if the OPCUA server is not available at the time of this connect attempt. Possibly further reconnections handle this?

Also here: Just continue without sucessfully connection makes the subsequent `defer client.Close()` inconsistent possibly and could cause the subsequent `processReadCommands` to fail.

Reply:
done